sdks/python/apache_beam/pipeline.py

```diff
@@ -808,7 +810,7 @@ class AppliedPTransform(object):
 
   def __init__(self,
                parent,
-               transform,  # type: ptransform.PTransform
+               transform,  # type: Optional[ptransform.PTransform]
```

Comment:
in `Pipeline.__init__`:

```python
    # Stack of transforms generated by nested apply() calls. The stack will
    # contain a root node as an enclosing (parent) node for top transforms.
    self.transforms_stack = [AppliedPTransform(None, None, '', None)]
```

Best way to deal with this may be a special `RootAppliedTransform` subclass.

---

I can also possibly be `None` in `AppliedPTransform.from_runner_api()`:

```python
    transform = ptransform.PTransform.from_runner_api(proto.spec, context)
    result = AppliedPTransform(
        parent=None,
        transform=transform,
        full_label=proto.unique_name,
        inputs=main_inputs)
```

This is because `PTransform.from_runner_api()` returns `Optional[PTransform]`

```python
  @classmethod
  def from_runner_api(cls,
                      proto,  # type: Optional[beam_runner_api_pb2.FunctionSpec]
                      context  # type: PipelineContext
                     ):
    # type: (...) -> Optional[PTransform]
    if proto is None or not proto.urn:
      return None
    parameter_type, constructor = cls._known_urns[proto.urn]
```

Reply:
On the second issue, I should point that mypy knows that `proto` is not `None` when we call `PTransform. from_runner_api ()` (because `proto.spec` is always non-None), and we could _almost_ use that knowledge to solve this problem with `@overload`s of `PTransform.from_runner_api()`, like this:

```python
  @overload
  @classmethod
  def from_runner_api(cls,
                      proto,  # type: None
                      context  # type: PipelineContext
                     ):
    # type: (...) -> None
    pass

  @overload
  @classmethod
  def from_runner_api(cls,
                      proto,  # type: beam_runner_api_pb2.FunctionSpec
                      context  # type: PipelineContext
                     ):
    # type: (...) -> PTransform
    pass
  
  @classmethod
  def from_runner_api(cls, proto, context):
    if proto is None or not proto.urn:
      return None
```

Unfortunately, typing can't track whether the value of `proto.urn` is empty.   Is there any chance that this could be changed to `proto.urn is None`?

Reply:
I don't think it'd ever be None, given that an unset proto field is the default value of that field.